src/graph.cpp

```diff
@@ -0,0 +1,107 @@
+//
+// Created by jmeintrup on 06.03.18.
+//
+
+#include <limits>
+#include "sealib/graph.h"
+
+/**
+ * Constructor for the graph, takes an adj_matrix as value in the following format:
+ * node x node matrix, entry in matrix is number of edges to that node.
+ *
+ * example:
+ *
+ * n         1       2       3
+ *      **************************
+ *  1   *    0   *   1   *   1   *
+ *      **************************
+ *  2   *    1   *   0   *   2   *
+ *      **************************
+ *  3   *    0   *   2   *   1   *
+ *      **************************
+ */
+Graph::Graph(unsigned int** adj_matrix, unsigned int _order):order(_order) {
+    nodes = std::vector<Node*>();
+
+    for(unsigned int i = 0; i < order; i ++) {
+        auto * n = new Node(adj_matrix[i], order);
```

Comment:
Hmm, is it possible to create a graph with nodes that have different number of neighbours?
In the current implementation it looks like every vertex has to be of the same degree.

Reply:
Underscores: Yes, they are necessary. Warning/Error on build otherwise, because of variable shadowing. CLion does not complain before build, though. 

It is possible to create graphs with nodes of different degree, the degree is calculated based on the entries of the adj_matrix[i] by summing up the values of that array.

Example array: [0,0,0,1,2] - summing up the entries gives us a degree of 3, the node would have one edge with node 3 and two edges with node 4 (countinng from 0).


Naming/Comments: Sure.

Reply:
Taking the graph that you use in your test and the graph constructor,
```cpp
adj_mtrx[0] = new unsigned int[order]{0,2,0,1};
adj_mtrx[1] = new unsigned int[order]{2,0,1,0};
adj_mtrx[2] = new unsigned int[order]{0,1,0,1};
adj_mtrx[3] = new unsigned int[order]{1,0,1,0};
```
all adjacency arrays must have the same length.

How this representation?

```cpp
adj_mtrx[0] = new unsigned int[2]{1, 2};
adj_mtrx[1] = new unsigned int[2]{0, 2};
adj_mtrx[2] = new unsigned int[3]{1, 3, 0};
adj_mtrx[3] = new unsigned int[1]{2};
```